mods/chrome-ui-tweaks.wh.cpp

```diff
@@ -0,0 +1,847 @@
+        (pfound = wcsstr(lpLibFileName, L"\\chrome.dll"),
+            pfound!=NULL && _wcsicmp(pfound, L"\\chrome.dll")==0) &&
+        !g_chromeDllLoaded.exchange(true)) {
+            Wh_Log(L"Loading chrome.dll from path: %s", lpLibFileName);
+            set_hooks(module);
+            Wh_ApplyHookOperations();
+    }
+    return module;
+}
+
+// The mod is being initialized, load settings, hook functions, and do other
+// initialization stuff if required.
+BOOL Wh_ModInit() {
+    // Check whether we are the main chrome process
+    bool mainProcess = true;
+    int argc;
+    LPWSTR* argv = CommandLineToArgvW(GetCommandLine(), &argc);
+    if (!argv) {
+        Wh_Log(L"CommandLineToArgvW failed");
+        return FALSE;
+    }
+
+    for (int i = 1; i < argc; i++) {
+        if (wcsncmp(argv[i], L"--type=", sizeof("--type=") - 1) == 0) {
+            mainProcess = false;
+            break;
+        }
+    }
+    LocalFree(argv);
+
+    if (!mainProcess) {
+        return FALSE;
+    }
+
+    // Initialize the mod for the main chrome process only
+    Wh_Log(L"Init " WH_MOD_ID L" version " WH_MOD_VERSION);
+
+    LoadSettings();
+
+    // if chrome.dll is already loaded by the process, we expect this to succeed
+    // otherwise assume we are during startup and it will be loaded later
+    HMODULE chromeModule = GetModuleHandleW(L"chrome.dll");
+    if (chromeModule == NULL) {
+        // hook LoadLibraryExW to apply our actual hooks later when chrome.dll is loaded
+        HMODULE kernelBaseModule = GetModuleHandle(L"kernelbase.dll");
+        FARPROC pKernelBaseLoadLibraryExW = GetProcAddress(kernelBaseModule, "LoadLibraryExW");
+
+        Wh_SetFunctionHook((void*)pKernelBaseLoadLibraryExW,
+                           (void*)LoadLibraryExW_Hook,
+                           (void**)&LoadLibraryExW_Original);
+        Wh_Log(L"Waiting for chrome.dll to be loaded");
+    } else {
+        set_hooks(chromeModule);
+    }
+
+    return TRUE;
+
+#pragma region "Symbol-based hooking (disabled)"
+/* Why don't we use symbol-based hooking in the first place?
+
+1. To search the symbols, we need to download chrome.dll.pdb from google's symbol server
+   for our chrome.dll version.
+2. chrome.dll.pdb is upwards of **4 GB**, which is all mapped into the process memory.
+3. over 50% of the time parsing the symbol table is spent in undecorating symbol names,
+   even with the noUndecoratedSymbols option set - maybe I just didn't figure out the
+   right way, or maybe undecorating is an unavoidable part of traversing symbols with
+   the DIA library Windhawk uses (msdia140_windhawk.dll).
+   The last time I tried it, it was busy parsing, undecorating, and traversing symbols
+   for over 4 hours (pegging one core at 100%) before I killed the process.
+4. While the symbol->address mapping can be cached, it needs to be updated every time
+   the program updates. Due to relatively frequent minor version and security updates,
```

Comment:
I thought about another possible approach: generating the mapping on the server. It will require some effort, but what I had in mind is having a repository with such mappings, and a scheduled GitHub Actions workflow that will run from time to time and check for Chrome updates. If there's an update, it will download it, get the symbols, and will created the mapping for the required functions. The mod will look for the mappings in that repository.

Pros:
* Advanced mods can be created more easily without having to create and maintain signatures.
* More resilient to Chrome updates.

Cons:
* Such a GitHub Actions workflow has to be developed.
* There can be a time gap until the mappings are updated to a newly released version (less of a problem IMO as Chrome updates gradually).
* Versions without a mapping (beta etc.) won't be supported.

I think what you did is perfectly fine, just some food for thought for future development.

In fact, I'll probably develop something like this, manual and temporary for now, due to [the Microsoft symbols incident](https://ramensoftware.com/windhawk-and-symbol-download-errors).

Reply:
I agree that sounds like a reasonable idea. I'm not familiar with Github Actions and probably won't try that myself, since I'm more focused on the mod itself (especially since it's already working). I'd probably want to switch over to that if you end up making it, maybe keeping the existing method as a fallback.